app/experimenter/nimbus-ui/src/hooks/useSearchParamsState.tsx

```diff
@@ -3,28 +3,49 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 import { useLocation, useNavigate } from "@reach/router";
-import { useEffect, useMemo } from "react";
+import React, {
+  createContext,
+  MutableRefObject,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+} from "react";
 
-export const genStorageKey = (subkey: string) => `searchParamState:${subkey}`;
+export type SearchParamStorage = Record<string, string>;
+export type SearchParamsContextState =
+  | MutableRefObject<SearchParamStorage>
+  | undefined;
 
-export function useSearchParamsState(storageSubKey?: string) {
+export const SearchParamsContext =
+  createContext<SearchParamsContextState>(undefined);
+
+export const SearchParamsStateProvider: React.FC = ({ children }) => {
+  const storage = useRef<SearchParamStorage>({});
+  return (
+    <SearchParamsContext.Provider value={storage}>
+      {children}
+    </SearchParamsContext.Provider>
+  );
+};
```

Comment:
An alternative I'd considered for this was to just define a context with a ref in the App component itself. But, I thought creating this hook-specific context provider would help contain the madness to this module and naturally scope the data stored in the ref.